### DISCUSSION TITLE
Problem flashing brand-new ESP32-CAM.

### DISCUSSION THREAD

**Discussion by laurentppol** [arendst/Tasmota#15693] [Q&A] [answered]
It is a bit OT (as this is not strictly Tasmota problem):
- I can't flash brand new ESP32-CAM (with small - also from Ali breadboard with USB),
- If I connect usinnnnng Minicom, I get strictly nothing, no reaction to 2 buttons (GPIO0 + RESET), 

- If I remove ESP from doughter board and reinsert I get:

`                                                                             
ets Jun  8 2016 00:22:57                                                     
                                                                             
rst:0x1 (POWERON_RESET),boot:0x3 (DOWNLOAD_BOOT(UART0/UART1/SDIO_REI_REO_V2))
waiting for download`

**If then try to upload firmware, I get success:**

`[laurent@laptop4 esptool]$ ./esptool.py --chip esp32 --before default_reset --after hard_reset --port /dev/ttyUSB5 --baud 921600 write_flash -z --flash_mode dout 0x0 /opt/esp-open-sdk.old/esptool/tasmota32-webcam.factory.bin -fm dout
esptool.py v2.3.1
Connecting....
Chip is ESP32D0WDQ6 (revision (unknown 0xa))
Features: WiFi, BT, Dual Core, VRef calibration in efuse
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 921600
Changed.
Configuring flash size...
Auto-detected Flash size: 4MB
Compressed 1294240 bytes to 865913...
Wrote 1294240 bytes (865913 compressed) at 0x00000000 in 11.0 seconds (effective 945.0 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
[laurent@laptop4 esptool]$ 
`

But returning to Minicom - still nothing. So doughterboard seems  to work. But something does not.
And remove-insert gives same reesult "waiting for download".
If I DON'T do the trick with reinserting module, but simply try to upload firmware second time I get timeout from esptool.pl.

Any ideas?

**Comment by sfromis** [comment]
For ESP32, you should use current esptool.py, v2.3.1 is ancient, but may have worked. However it saying `(revision (unknown 0xa))` is a bad sign.

After flashing, you should disconnect gpio0 from ground, and power cycle the device to get out of flashing mode. If everything is ok, then it should boot normally, and broadcast a Wifi AP to connect to, and of course also have a serial console available. Note that many computer USB sockets may be too low in power for ESP32 to operate reliably, especially with camera. Better to use a stronger power supply, at least 1A of 5V.

BTW: My experience with those cheap ESP32-CAM modules is that quality may be "varying", to put it nicely.

**Reply by laurentppol** [reply]
But using same tool, same HW (USB in laptop) I have successfully flashed ESP32-DevKitC.
Where (now) is ESP-IDF located? Still on espressif WWW?

Removed cam module, same thing. I CAN put ESP into flash mode using GPIO0 button. Without board removal "trick".
But still nothing.
Have You remarked "Hash of data verified"?

Found ESP-IDF, but I am on backup (3G) link, so don't know if download will be succesful. If not will retry as main link comes up.
I think I have 3 such modules, at least one (don't know if I find this) I have programmed OK, but at time Tasmota-CAM was beta.

**Reply by sfromis** [reply]
You wrote "I can't flash", and also can. And flashing a dev board is not same HW, as you're likely not using a separate USB/serial adapter then. Anyway, in case of borderline power situation, do not expect that "worked with one board" means that it should work with all boards. And do drop that **ancient** `esptool.py`, your output shows that it does not work perfectly.

AFAIK, https://github.com/Jason2866/esp-idf is used to have a verified ESP IDF for Tasmota, sometimes tweaked a bit.

Anyway, my "BTW" situation includes "surprising difficulties" with getting successful flash of those *crappy* boards.

**Reply by laurentppol** [reply]
Just trying to download v4.2 (github of espressif). So dropping :)
What gives Jason's IDF for normal user?

**Reply by sfromis** [reply]
What I meant by "dropping" was not related to ESP-IDF, but about getting a current [esptool.py](https://github.com/espressif/esptool/) where version 4.0.1 was released a short time ago.

The Jason2866 fork of ESP-IDF is not so much meant to be used outside of Tasmota, more to have a stable and tested version "confirmed" to work well with Tasmota. AFAIK the main difference is selection of which patches to avoid, and which to integrate "ahead of time". 

But to be clear, I was in no way making a suggestion of even considering picking a different ESP-IDF. Since you asked about ESP-IDF, I answered, without any assumptions about why you suddenly asked about something unrelated.

**Reply by laurentppol** [reply]
esptool.py is a part of ESP-IDF? Isn't it?

BTW, I really appreciate Your work with Tasmota. This leaves me free (no need to deevelop own firmmware) to work :)

**Reply by sfromis** [reply]
No, esptool.py is a narrow tool specifically for flashing (I provided a link), while ESP-IDF is a much broader development framework, basically for producing the firmware to be flashed. From the [docs](https://github.com/espressif/esp-idf#flashing-the-project) for ESP-IDF:
> When the build finishes, it will print a command line to use esptool.py to flash the chip.

Hence esptool.py is not part of ESP-IDF, but expected to be used by the developer after the build finished. I'd not be surprised if both tools are often installed at the same time, without the user doing a separate install of esptool.py, but that's just my speculation.

For context: ESP-IDF is for ESP32 chips, while RTOS SDK is their framework for ESP8266. And `esptool.py` serves both architectures, is part of neither tool, and was initially created outside of Espressif, which is supporting it now.

**Reply by laurentppol** [reply]
Ok: 2 next tests:
- downloaded Espressif v4.4, esptool v.3.x,       
- connected ESPCAM to an USB hub, powered from 5V/3A PS.

Same result. Except ESP chip identified.
Next idea? (apart throwing this ESP32?).

Ask on Espressif forum? Maybe there is someone knowing essptool.py internals?

**Reply by ascillato** [reply]
Have you tried the official Tasmota's web-installer for ESP32CAM that is at https://tasmota.github.io/docs/Getting-Started/ ?

If I were you, I will start there since the web-installer takes care of all files and flashing process. And if that works, then you can move to esptool.py, if you really want to use it.

The ESP32CAM is a very cheap hardware with poor design but it works fine. I use several of those and the unique problem I had was when the power supply broke down after 2 years of installed and burnt the cam's chip due to overvoltage. The ESP32CAM gets hot very fast, but with proper ventilation it has no problems. For latest ones I flashed last month, I used the web-installer and it worked just fine. No need to worry about the files nor the esptool.py.

**Comment by laurentppol** [comment, marked as answer]
Finally GOT IT:
- not 100% sure, but key was to ERASE_FLASH before programming. And playing with power / GPIO0 (RESET button seems have no effect).
Anyway have 2 modules with Tasmota32-webcam.factory.bin flashed. On one module it seems camera went dead, but this       is not a problem.

Question [OT] - why firmware allways tries to init camera? Even if board type is set to ESP32-DevKitC? Should I reflash to "regular" Tasmota32 (if no plans to use camera on one of modules, but want I2C, ADS1115 and INA226 - for the last one I think I'll have to compile my own like for ESP8266?)
Where to find which pins I have free to use on WebCAM module (camera removed)?

**Reply by sfromis** [reply]
Yes, if you do not want to use the camera, better upgrade to regular tasmota32.bin which includes a lot of sensor support already. If the camera is not connected, I'd expect no restrictions on what pins you can use, unless you intend to use the SD card, of course. The onboard LED is connected to gpio 4.

**Reply by sfromis** [reply]
Default built-in template should be: 
![image](https://user-images.githubusercontent.com/47082390/170558344-626f493d-191f-40f3-9a00-e8de0430068d.png)

**Comment by JohnyDNU** [comment]
_Found this in Google while trying to resolve the same issue, so describe my solution if anyone else need a help._

Well, I spent a lot of time to flash my ESP32-CAM, and finnaly I found the only working solution.

The main idea: **are you sure your board is in bootload state**? :-)

That is not enough to ground GPIO0 and power up the board. You need to **press** the small **hardware button Reset** on the board's chip side after power up!

I just connect +5V, RX, TX and GND to my USB dongle, connect GPIO0 to GND, plug into USB port and press Reset button. Now it runs the bootloader and ready to be flasshed even through browser!
